### PR TITLE
Autoselect location from user data

### PR DESCRIPTION
**Pull request by esoergel**
Part of [this spec](https://docs.google.com/document/d/1Pji_1XJSJUMmI4wvqKWUJ77AerqkrB-dtardcN60FO4/edit#heading=h.q3wyha58n4s0)
@snopoke
Looking for feedback at this time.  I'm pretty unfamiliar with the app builder code, so if there's a better place for me to insert this, lemme know.  I'm also not crazy about adding this only if `app.commtrack_enabled` - it should depend on `domain.commtrack_enabled`.  Can we remove that checkbox from the UI and make all CommTrack domains set that flag to True?  It looks like the suite generator doesn't rely on the concept of a domain, so I'd like to avoid introducing that if not necessary.

### CODE REVIEW

File: corehq/apps/app_manager/suite_xml.py [hunk 1]
```diff
@@ -1416,6 +1416,22 @@ def add_referenced_instances(entry, details_by_id, relevance_by_menu, menu_by_co
 
         entry.require_instance(*instances)
 
+    def get_location_autoselect(self):
+        from .models import AUTO_SELECT_LOCATION
+        user_data_key = 'commtrack-supply-point'
+        xpath = XPath(u"instance('commcaresession')"
+                       "/session/user/data/{}".format(user_data_key))
```
Comment (dimagimon) [inline corehq/apps/app_manager/suite_xml.py]: E127 continuation line over-indented for visual indent

Comment (snopoke) [inline corehq/apps/app_manager/suite_xml.py]: you should use `session_var(user_data_key, path='user/data')`

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -365,7 +366,8 @@ class AutoSelectCase(DocumentSchema):
                         xpath expression.
 
     """
-    mode = StringProperty(choices=[AUTO_SELECT_USER, AUTO_SELECT_FIXTURE, AUTO_SELECT_CASE, AUTO_SELECT_RAW])
+    mode = StringProperty(choices=[AUTO_SELECT_USER, AUTO_SELECT_FIXTURE,
+            AUTO_SELECT_CASE, AUTO_SELECT_LOCATION, AUTO_SELECT_RAW])
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py]: probably don't want to add this here (or we should also update the UI for advanced forms) but for advanced forms you can do the autoload manually via a user data so I don't think we need it.

File: corehq/apps/app_manager/suite_xml.py [hunk 3]
```diff
@@ -1816,6 +1838,7 @@ def get_target_module(case_type, module_id, with_product_details=False):
                 })
 
         if module.get_app().commtrack_enabled:
+            # TODO add here?
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml.py]: Don't think we need anything here (similar comment to above about advanced forms)

File: corehq/apps/app_manager/suite_xml.py [hunk 4]
```diff
@@ -1416,6 +1416,22 @@ def add_referenced_instances(entry, details_by_id, relevance_by_menu, menu_by_co
 
         entry.require_instance(*instances)
 
+    def get_location_autoselect(self):
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml.py]: This is basically the same code as here: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/suite_xml.py#L1682-1688

Might make sense to make the function usable in both places (`def get_userdata_autoselect` or something)

**Comment by snopoke** [comment]
Generally looks good. On the comment of `app.commtrack_enabled` happy to kill that setting and just use the domain value if we can get that into the right places.

File: corehq/apps/app_manager/models.py (hunk 2 above)
Comment (esoergel) [inline corehq/apps/app_manager/models.py]: :+1: I'll remove it

**Comment by esoergel** [comment]
Do you have suggestions on how to pass domain in appropriately?  (it may even be to just make that enabled on all apps on commtrack domains).  I don't want to break the abstraction if it's avoidable.

**Comment by snopoke** [comment]
You can either pass it in when creating the SuiteGenerator or we can make
sure the commtrack field gets set on all apps in commcare supply domains or
create a utility function that you can call that checks the status. I don't
have an answer of which is better.

On 20 May 2015 at 01:35, Ethan Soergel <notifications@github.com> wrote:

> Do you have suggestions on how to pass domain in appropriately? (it may
> even be to just make that enabled on all apps on commtrack domains). I
> don't want to break the abstraction if it's avoidable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/6743#issuecomment-103697806>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

File: corehq/apps/app_manager/tests/data/suite/suite-advanced-commtrack.xml [hunk 5]
```diff
@@ -294,9 +294,26 @@
         <locale id="forms.m0f0"/>
       </text>
     </command>
+    <instance id="casedb" src="jr://instance/casedb"/>
+    <instance id="commcaresession" src="jr://instance/session"/>
```
Comment (esoergel) [inline corehq/apps/app_manager/tests/data/suite/suite-advanced-commtrack.xml]: I had to add in these nodes for the commtrack suite, as this is default behavior for all apps on a commtrack domain.

File: corehq/apps/app_manager/models.py [hunk 6]
```diff
@@ -3656,10 +3657,17 @@ class Application(ApplicationBase, TranslationMixin, HQMediaMixin):
...
     commtrack_requisition_mode = StringProperty(choices=CT_REQUISITION_MODES)
     auto_gps_capture = BooleanProperty(default=False)
 
+    @property
+    @memoized
+    def commtrack_enabled(self):
+        if settings.UNIT_TESTING:
+            return False  # override with .tests.util.commtrack_enabled
+        domain_obj = Domain.get_by_name(self.domain) if self.domain else None
+        return domain_obj.commtrack_enabled if domain_obj else False
```
Comment (esoergel) [inline corehq/apps/app_manager/models.py]: I spoke with @dannyroberts about pulling the lookup out to a separate `get_domain_context` function which handles the environment-specific lookup, but it seemed a bit overblown for this one usage.  If there arise more places that would like context from the domain object, that does seem like a good way to go about it.

File: corehq/apps/app_manager/suite_xml.py [hunk 7]
```diff
@@ -1421,6 +1421,12 @@ def add_referenced_instances(entry, details_by_id, relevance_by_menu, menu_by_co
 
         entry.require_instance(*instances)
 
+    def get_userdata_autoselect(self, key, session_id, mode):
+        xpath = session_var(key, path='user/data')
+        datum = SessionDatum(id=session_id, function=xpath)
+        assertions = self.get_auto_select_assertions(xpath, mode, [key])
+        return datum, assertions
```
Comment (snopoke) [inline corehq/apps/app_manager/suite_xml.py]: don't think this should be here. Bad merge>

Comment (snopoke) [inline corehq/apps/app_manager/suite_xml.py]: never mind, my bad

File: corehq/apps/app_manager/views.py [hunk 8]
```diff
@@ -691,7 +691,7 @@ def get_apps_base_context(request, domain, app):
                                and toggles.APP_BUILDER_CAREPLAN.enabled(request.user.username)),
             'show_advanced': (v2_app
                                and (toggles.APP_BUILDER_ADVANCED.enabled(request.user.username)
-                                    or getattr(app, 'commtrack_enabled', False))),
+                                    or app.commtrack_enabled)),
```
Comment (snopoke) [inline corehq/apps/app_manager/views.py]: I think this still needs to use `getattr` since `RemoteApps` don't have a `commtrack_enabled` property.

**Comment by snopoke** [comment]
kicked build

**Comment by esoergel** [comment]
Updated.  The previous build passed, the only change was switching back to `getattr`

**Pull request merged**